File: problems/39/solution.java
// Bad solution. Need to re-write

// TreeNode type
// class TreeNode {
//     int val;
//     TreeNode left;
//     TreeNode right;
//     TreeNode(int x) { val = x; left = null; right = null; }
// }
class Solution {
    ArrayList<Integer> result = new ArrayList<Integer>();

    public ArrayList<Integer> inorderTraversal(TreeNode root) {
        // Implement the solution
        _inorder(root);
        return result;
    }


    public void _inorder(TreeNode root) {
        if (root == null)
            return;

        _inorder(root.left);
        result.add(root.val);
        _inorder(root.right);
    }

    private int hasTwoSum(int K) {
        ArrayList<Integer> order = new ArrayList<Integer>();
        order = result;
        if (order.size() <= 1) {
            return 0;
        }
        int start = 0;
        int end = order.size() - 1;
        while (start < end) {
            if (order.get(start) + order.get(end) == K)
                return 1;
            if (order.get(start) + order.get(end) > K)
                end --;
            else
                start++;
        }
        return 0;
    }

    public int isPairPresent(TreeNode root, int K) {
        if (root == null)
            return 0;
        _inorder(root);
        return hasTwoSum(K);
    }
}

File: problems/39/solution.cpp
// 	 struct TreeNode {
// 	     string val;
// 	     TreeNode* left;
//       TreeNode* right;
//       TreeNode(string x): val(x), left(NULL), right(NULL) {}
//   };
//
//   class Solution {
// 	   public:
// 		  int isPairPresent(TreeNode* tree, int K);
//   };
//
//   Please refrain from using print statements

void inorder(TreeNode* root, vector<int>& order) {
    if (!root)
        return;
    inorder(root->left, order);
    order.push_back(root->val);
    inorder(root->right, order);
}

int hasTwoSum(const vector<int>& order, int K) {
    if (order.size() <= 1)
        return 0;
    int start = 0, end = (int)order.size() - 1;
    while (start < end) {
        if (order[start] + order[end] == K)
            return 1;
        if (order[start] + order[end] > K)
            end --;
        else
            start++;
    }
    return 0;
}

int Solution::isPairPresent(TreeNode* root, int K) {
    if (!root) {
        return 0;
    }

    vector<int> order;
	inorder(root, order);
	return hasTwoSum(order, K);
}
